source/codegen/templates/service_helpers.mako

```diff
@@ -271,6 +271,227 @@ ${populate_response(function_data=function_data, parameters=parameters)}\
+  streaming_param_to_include = common_helpers.get_input_streaming_param(parameters)
+%>\
+  struct ${struct_name}
+  {
+     % for param in streaming_param_to_include:
+     ${param['type']} ${param['name']};
+     % endfor
+     ${service_class_prefix.lower()}_grpc::${grpc_streaming_type} data;
+     std::shared_ptr<${service_class_prefix}LibraryInterface> library;
+  };
+</%def>
+
+<%def name="define_moniker_function_body(function_name, function_data)">\
+<%
+  config = data['config']
+  struct_name = common_helpers.get_data_moniker_struct_name(function_name)
+  moniker_function_name = common_helpers.get_data_moniker_function_name(function_name)
+  streaming_param = common_helpers.get_streaming_parameter(function_data['parameters'])
+  streaming_type = streaming_param['type']
+  grpc_streaming_type = streaming_param['grpc_streaming_type']
+  arg_string = service_helpers.create_args(function_data['parameters'])
+  arg_string = arg_string.replace(", &moniker", "").strip()
+  data_type = streaming_type.replace("[]", "")
+  c_api_name = service_helpers.get_c_api_name(function_name)
+  streaming_param_to_include = common_helpers.get_input_streaming_param(function_data['parameters'])
+%>\
+::grpc::Status ${moniker_function_name}(void* data, google::protobuf::Arena& arena, google::protobuf::Any& packedData)
+{
+    ${struct_name}* function_data = (${struct_name}*)data;
+    auto library = function_data->library;
+    ${initialize_moniker_streaming_parameters(streaming_param_to_include)}\
+    % if streaming_param and streaming_param['direction'] == 'out':
+        ${streaming_handle_out_direction(c_api_name, arg_string, data_type, streaming_type, streaming_param)}\
+    % elif streaming_param and streaming_param['direction'] == 'in':
+        ${streaming_handle_in_direction(c_api_name, arg_string, data_type, grpc_streaming_type, streaming_type, streaming_param)}\
+    % endif
+    if (status < 0) {
+      std::cout << "${moniker_function_name} error: " << status << std::endl;
+    }
+    return ::grpc::Status::OK;
+}
+</%def>
+
+<%def name="define_streaming_api_body(function_name, function_data, parameters)">\
+<%
+  config = data['config']
+  output_params = [p for p in parameters if common_helpers.is_output_parameter(p)]
+  service_class_prefix = config["service_class_prefix"]
+  request_param = service_helpers.get_request_param(function_name)
+  response_param = service_helpers.get_response_param(function_name)
+  struct_name = common_helpers.get_data_moniker_struct_name(function_name)
+  moniker_function_name = common_helpers.get_data_moniker_function_name(function_name)
+  streaming_param = common_helpers.get_streaming_parameter(parameters)
+  streaming_param_to_include = common_helpers.get_input_streaming_param(parameters)
+%>\
+${initialize_streaming_input_param(function_name, streaming_param_to_include, parameters, streaming_param)}
+      ${struct_name}* data = new ${struct_name}();
```

Comment:
More raw `new`s here. Can some of these be allocated on the stack? Or can they be a `std::unique` pointers?

There is one case where we may "need" raw pointers: if we are passing the data to asynchronous code that is not in our control. The data needs to be kept around after this function returns (cannot be a stack variable) and the other code may not support managing data with `unique_ptr`s (or other smart pointers).

 But we need to be clear when that is happening and what the cleanup strategy is.

Reply:
Updated it to use unique_ptr.